File: frontend/src/app/src/app/services/dance.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { DanceCategory } from '../models/dance-category.model';
import { DanceSchool } from '../models/dance-school.model';
import { Course } from '../models/course.model';

@Injectable({
  providedIn: 'root'
})
export class DanceService {

  private apiUrl = 'http://localhost:8080/api'; // Remplacez par l'URL de votre API Spring Boot

  constructor(private http: HttpClient) { }

  // -------------------- DanceCategory CRUD --------------------
  
  // Créer une DanceCategory
  createDanceCategory(category: DanceCategory): Observable<DanceCategory> {
    return this.http.post<DanceCategory>(`${this.apiUrl}/categories`, category);
  }

  // Obtenir toutes les DanceCategories
  getAllDanceCategories(): Observable<DanceCategory[]> {
    return this.http.get<DanceCategory[]>(`${this.apiUrl}/categories`);
  }

  // Obtenir une DanceCategory par ID
  getDanceCategoryById(id: number): Observable<DanceCategory> {
    return this.http.get<DanceCategory>(`${this.apiUrl}/categories/${id}`);
  }

  // Mettre à jour une DanceCategory
  updateDanceCategory(id: number, updatedCategory: DanceCategory): Observable<DanceCategory> {
    return this.http.put<DanceCategory>(`${this.apiUrl}/categories/${id}`, updatedCategory);
  }

  // Supprimer une DanceCategory
  deleteDanceCategory(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/categories/${id}`);
  }

  // -------------------- DanceSchool CRUD --------------------
  
  // Créer une DanceSchool
  createDanceSchool(school: DanceSchool): Observable<DanceSchool> {
    return this.http.post<DanceSchool>(`${this.apiUrl}/schools`, school);
  }

  // Obtenir toutes les DanceSchools
  getAllDanceSchools(): Observable<DanceSchool[]> {
    return this.http.get<DanceSchool[]>(`${this.apiUrl}/schools`);
  }

  // Obtenir une DanceSchool par ID
  getDanceSchoolById(id: number): Observable<DanceSchool> {
    return this.http.get<DanceSchool>(`${this.apiUrl}/schools/${id}`);
  }

  // Mettre à jour une DanceSchool
  updateDanceSchool(id: number, updatedSchool: DanceSchool): Observable<DanceSchool> {
    return this.http.put<DanceSchool>(`${this.apiUrl}/schools/${id}`, updatedSchool);
  }

  // Supprimer une DanceSchool
  deleteDanceSchool(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/schools/${id}`);
  }

  // -------------------- Course CRUD --------------------
  
  // Créer un Course
  createCourse(course: Course): Observable<Course> {
    return this.http.post<Course>(`${this.apiUrl}/courses`, course);
  }

  // Obtenir tous les Courses
  getAllCourses(): Observable<Course[]> {
    return this.http.get<Course[]>(`${this.apiUrl}/courses`);
  }

  // Obtenir un Course par ID
  getCourseById(id: number): Observable<Course> {
    return this.http.get<Course>(`${this.apiUrl}/courses/${id}`);
  }

  // Mettre à jour un Course
  updateCourse(id: number, updatedCourse: Course): Observable<Course> {
    return this.http.put<Course>(`${this.apiUrl}/courses/${id}`, updatedCourse);
  }

  // Supprimer un Course
  deleteCourse(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/courses/${id}`);
  }
}

File: dancebackend/src/main/java/com/dancebackend/dancebackend/service/DanceService.java
package com.dancebackend.dancebackend.service;

import com.dancebackend.dancebackend.entity.Course;
import com.dancebackend.dancebackend.entity.DanceCategory;
import com.dancebackend.dancebackend.entity.DanceSchool;
import com.dancebackend.dancebackend.repository.CourseRepository;
import com.dancebackend.dancebackend.repository.DanceCategoryRepository;
import com.dancebackend.dancebackend.repository.DanceSchoolRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class DanceService {

    private final DanceCategoryRepository danceCategoryRepository;
    private final DanceSchoolRepository danceSchoolRepository;
    private final CourseRepository courseRepository;

    public DanceService(DanceCategoryRepository danceCategoryRepository, DanceSchoolRepository danceSchoolRepository, CourseRepository courseRepository) {
        this.danceCategoryRepository = danceCategoryRepository;
        this.danceSchoolRepository = danceSchoolRepository;
        this.courseRepository = courseRepository;
    }

    // -------------------- DanceCategory CRUD --------------------

    public DanceCategory createDanceCategory(DanceCategory category) {
        return danceCategoryRepository.save(category);
    }

    public List<DanceCategory> getAllDanceCategories() {
        return danceCategoryRepository.findAll();
    }

    public Optional<DanceCategory> getDanceCategoryById(Long id) {
        return danceCategoryRepository.findById(id);
    }

    public DanceCategory updateDanceCategory(Long id, DanceCategory updatedCategory) {
        return danceCategoryRepository.findById(id).map(category -> {
            category.setName(updatedCategory.getName());
            return danceCategoryRepository.save(category);
        }).orElseThrow(() -> new RuntimeException("DanceCategory not found with id: " + id));
    }

    public void deleteDanceCategory(Long id) {
        danceCategoryRepository.deleteById(id);
    }

    // -------------------- DanceSchool CRUD --------------------

    public DanceSchool createDanceSchool(DanceSchool school) {
        return danceSchoolRepository.save(school);
    }

    public List<DanceSchool> getAllDanceSchools() {
        return danceSchoolRepository.findAll();
    }

    public Optional<DanceSchool> getDanceSchoolById(Long id) {
        return danceSchoolRepository.findById(id);
    }

    public DanceSchool updateDanceSchool(Long id, DanceSchool updatedSchool) {
        return danceSchoolRepository.findById(id).map(school -> {
            school.setName(updatedSchool.getName());
            school.setLocalisation(updatedSchool.getLocalisation());
            school.setOpenday(updatedSchool.getOpenday());
            school.setCloseday(updatedSchool.getCloseday());
            school.setOpentime(updatedSchool.getOpentime());
            school.setClosetime(updatedSchool.getClosetime());
            return danceSchoolRepository.save(school);
        }).orElseThrow(() -> new RuntimeException("DanceSchool not found with id: " + id));
    }

    public void deleteDanceSchool(Long id) {
        danceSchoolRepository.deleteById(id);
    }

    // -------------------- Course CRUD --------------------

    public Course createCourse(Course course) {
        return courseRepository.save(course);
    }

    public List<Course> getAllCourses() {
        return courseRepository.findAll();
    }

    public Optional<Course> getCourseById(Long id) {
        return courseRepository.findById(id);
    }

    public Course updateCourse(Long id, Course updatedCourse) {
        return courseRepository.findById(id).map(course -> {
            course.setTitle(updatedCourse.getTitle());
            course.setContenu(updatedCourse.getContenu());
            course.setInstructor(updatedCourse.getInstructor());
            return courseRepository.save(course);
        }).orElseThrow(() -> new RuntimeException("Course not found with id: " + id));
    }

    public void deleteCourse(Long id) {
        courseRepository.deleteById(id);
    }
}